guava/src/com/google/common/base/Optional.java

```diff
@@ -282,7 +284,7 @@
    * @throws NullPointerException if the function returns {@code null}
    * @since 12.0
    */
-  public abstract <V> Optional<V> transform(Function<? super T, V> function);
+  public abstract <V extends Object> Optional<V> transform(Function<? super T, V> function);
```

Comment:
Your version is equivalent to `<V extends @NonNull Object>`, and is therefore correct.
However, I suggest that you include the qualifier, for emphasis.  Otherwise, someone who is reading the code might think that the `extends Object` is irrelevant (which it is in plain Java), and some IDEs will suggest removing it.